dspace-api/src/main/java/org/dspace/google/GoogleAsyncEventListener.java

```diff
@@ -0,0 +1,236 @@
+                            ue.getObject().getType() == Constants.BITSTREAM) {
+
+                        // Client Id, should uniquely identify the user or device. If we have a session id for the user
+                        // then lets use it, else generate a UUID.
+                        String cid;
+                        if (ue.getRequest().getSession(false) != null) {
+                            cid = ue.getRequest().getSession().getId();
+                        } else {
+                            cid = UUID.randomUUID().toString();
+                        }
+                        buffer.add(new GoogleAnalyticsEvent(cid, getIPAddress(ue.getRequest()), ue.getRequest()
+                                .getHeader("USER-AGENT"), ue.getRequest().getHeader("referer"), ue.getRequest()
+                                .getRequestURI() + "?" + ue.getRequest().getQueryString(),
+                                getObjectName(ue), System.currentTimeMillis()));
+                    }
+                } catch (Exception e) {
+                    log.error("Failed to add event to buffer", e);
+                    log.error("Event information: " + ue);
+                    Context context = ue.getContext();
+                    if (context != null) {
+                        log.error("Context information:");
+                        log.error("    Current User: " + context.getCurrentUser());
+                        log.error("    Extra log info: " + context.getExtraLogInfo());
+                        if (context.getEvents() != null && !context.getEvents().isEmpty()) {
+                            for (int x = 1; x <= context.getEvents().size(); x++) {
+                                log.error("    Context Event " + x + ": " + context.getEvents().get(x));
+                            }
+                        }
+                    } else {
+                        log.error("UsageEvent has no Context object");
+                    }
+                }
+            }
+        }
+    }
+
+    private String getObjectName(UsageEvent ue) {
+        try {
+            if (ue.getObject().getType() == Constants.BITSTREAM) {
+                // For a bitstream download we really want to know the title of the owning item
+                // rather than the bitstream name.
+                return ContentServiceFactory.getInstance().getDSpaceObjectService(ue.getObject())
+                        .getParentObject(ue.getContext(), ue.getObject()).getName();
+            } else {
+                return ue.getObject().getName();
+            }
+        } catch (SQLException e) {
+            // This shouldn't merit interrupting the user's transaction so log the error and continue.
+            log.error("Error in Google Analytics recording - can't determine ParentObjectName for bitstream " +
+                    ue.getObject().getID(), e);
+        }
+
+        return null;
+
+    }
+
+    private String getIPAddress(HttpServletRequest request) {
```

Comment:
This method seems like the same general logic as `ClientInfoService.getClientIp(request)`.  We might want to remove this method and replace it with a call to the existing method.